### PR TITLE
ICNR inizialization

### PR DESCRIPTION
**Pull request by cattaneod**
ICNR inizialization from "Checkerboard artifact free sub-pixel convolution" - Andrew Aitken et al. (2017) - https://arxiv.org/abs/1707.02937

This inizialization should be used in the last Conv2d layer before PixelShuffle

### CODE REVIEW

**Comment by ezyang** [comment]
@apaszke Do we want this one?

@catta202000  We can't merge this without a test but let us first decide if we want to put this in core.

**Comment by apaszke** [comment]
Hmm, not sure. It's relatively new, so I'd be reluctant to add it unless there have been some other papers that used it. Also, we should at least make it `icnr` to match the conventions used in other methods. I'll leave the decision up to @soumith

**Comment by makslevental** [comment]
@catta202000 can you explain please why you have to do all of the transpose/view tricks? isn't the point to simply replicate 1 subkernel across all convolution out channels?

**Comment by catta202000** [comment]
@makslevental yes, the basic idea is to copy the same subkernel across all the convolutions.
However, if you have multiple classes, you want to have different subkernels.

For example, if you have 5 classes, my implementation first generates 5 different subkernels, and then copy them accordingly to how PixelShuffle works.

**Comment by f0k** [comment]
If this ever gets merged, it should support both subpixel convolution and transposed convolution (which are different implementations of the same operation, but need a different arrangement for the kernel).

**Comment by ashnair1** [comment]
Any updates on this?

**Comment by f0k** [comment]
> Any updates on this?

FWIW, I coded up an inplace initializer for strided (transposed) convolutions as follows:
```python
def conv_nn_resize_(tensor_or_module, subinit, stride=None):
    """
    Nearest neighbor resize initialization for strided convolution.
    https://arxiv.org/abs/1707.02937
    """
    if isinstance(tensor_or_module, nn.Module):
        tensor = tensor_or_module.weight.data
        stride = stride or tensor_or_module.stride
    else:
        tensor = tensor_or_module
    try:
        stride = tuple(stride)
    except TypeError:
        stride = (stride,) * (tensor.dim() - 2)
    if all(s == 1 for s in stride):
        return subinit(tensor)
    subtensor = tensor[(slice(None), slice(None)) +
                       tuple(slice(None, None, s) for s in stride)]
    subinit(subtensor)
    for d, s in enumerate(stride):
        subtensor = subtensor.repeat_interleave(s, 2 + d)
    return tensor.copy_(subtensor[tuple(slice(None, l) for l in tensor.shape)])
```
You can either apply it directly to a convolution or transposed convolution module (no matter which dimensionality), or to a weight tensor (then you need to specify the stride). The second argument must be another inplace initializer such as `torch.nn.init.kaiming_uniform_` (or `functools.partial(nn.init.kaiming_uniform_, a=0.01)` if you need to pass arguments to it).

Note that this is for standard strided convolutions and strided transposed convolutions, not for combinations with the pixel shuffle layer (which is generally [not needed](https://arxiv.org/abs/1609.07009)).

**Comment by xunhuang1995** [comment]
@f0k Why is pixel shuffle layer generally not needed? I thought the paper you cite argues that convolution + pixel shuffle is better than transposed convolution?

**Comment by f0k** [comment]
> @f0k Why is pixel shuffle layer generally not needed? I thought the paper you cite argues that convolution + pixel shuffle is better than transposed convolution?

No, if I didn't misunderstand, they argue that convolution + pixel shuffle (= convolution in LR space) is better than upsampling + convolution (= convolution in HR space).

convolution + pixel shuffle is the same as a transposed convolution with a kernel size divisible by the stride (this is important, as demonstrated in https://distill.pub/2016/deconv-checkerboard/). I haven't understood *why* they used pixel shuffling instead of transposed convolutions, but it's possible that this was faster back then (cuDNN was much less mature).
